examples/threshold_sig.rs

```diff
@@ -0,0 +1,203 @@
+    fn step(&mut self) {
+        if let Some(block) = self.run_consensus() {
+            self.chat_log.push(block);
+        }
+    }
+
+    // Our chat protocol's consensus algorithm. Produces a new block to be appended to the chat
+    // log. Our consensus uses threshold-signing to verify that a message has received enough
+    // signature shares (i.e. has been signed by `threshold + 1` nodes).
+    fn run_consensus(&self) -> Option<(UserId, Msg, Signature)> {
+        // Create a new `MsgDatabase` of every message that has been signed by a validator node.
+        let all_pending: MsgDatabase = self.nodes.iter()
+            .fold(BTreeMap::new(), |mut all_pending, node| {
+                for (user_id, signed_msgs) in node.pending.iter() {
+                    let mut user_msgs = all_pending
+                        .entry(*user_id)
+                        .or_insert(BTreeMap::new());
+                    for (msg, sigs) in signed_msgs.iter() {
+                        let sigs: Vec<NodeSignature> = sigs.iter().map(|sig| sig.clone()).collect();
+                        user_msgs
+                            .entry(msg.to_string())
+                            .or_insert(Vec::new())
+                            .extend(sigs);
+                    }
+                }
+                all_pending
+            });
+
+        // Iterate over the `MsgDatabase` numerically by user ID, then iterator over each user's
+        // messages alphabetically. Try to combine the validator signatures. The first message that
+        // has received `threshold + 1` node signatures, will produce a valid "combined" signature
+        // and will be added to the chat log.
+        for (user_id, signed_msgs) in all_pending.iter() {
+            for (msg, sigs) in signed_msgs.iter() {
+                let sigs: Vec<(NodeId, &SignatureShare)> = sigs.iter()
+                    .map(|node_sig| (node_sig.node_id, &node_sig.sig))
+                    .collect();
+                if let Ok(sig) = self.pk_set.combine_signatures(sigs) {
+                    return Some((*user_id, msg.clone(), sig));
+                }
+            }
+        }
+
+        None
+    }
+}
+
+// A node the network that is running our chat protocol.
+struct Node {
+    id: NodeId,
+    sk_share: SecretKeyShare,
+    pending: MsgDatabase,
+}
+
+impl Node {
+    fn new(id: NodeId, sk_share: SecretKeyShare) -> Self {
+        Node { id, sk_share, pending: BTreeMap::new() }        
+    }
+
+    // Receives a message from a user, signs with message with the node's signing-key share, then
+    // adds the signed message to its database of `pending` messages.
+    fn recv(&mut self, user_id: UserId, msg: Msg) {
+        let sig = NodeSignature {
+            node_id: self.id,
+            sig: self.sk_share.sign(msg.as_bytes()),
+        };
+        self.pending
+            .entry(user_id)
+            .or_insert(BTreeMap::new())
+            .entry(msg)
+            .or_insert(Vec::new())
```

Comment:
`.or_insert_with(Vec::new)` is preferable, since it instantiates a `Vec` only if the entry is actually absent. (Also elsewhere.)